### PR TITLE
Compile time builder

### PR DESCRIPTION
**Pull request by ltratt**
This PR adds a simple compile-time builder to lrlex. Basically, if you want to hard-code a specific '.l' file(s) into your project, this gives you a way of doing that (much as traditional `lex` generates `.c` files). Admittedly, in one way this isn't very useful, at least not for performance (lrlex analyses `.l` files so fast that it's not worth worrying about) but a) it matches traditional expectations b) it means you don't have to cart around `.l` files separate from the binary. The `.rs` files that are created are naive, but, for the time being, probably sufficient. The interface that they expose seems simple enough that I think it'll stick.

To see this in action, see https://github.com/softdevteam/lrlex/pull/22/commits/14c65bbbc4dfd9abc7d5273f90f6cc546b97c0da (which only really makes sense when you build it locally IMHO), which adds a tiny calculator lexing file using this framework.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,2 +1,5 @@
 language: rust
 rust: nightly
+script:
+    - cargo test
+    - cd examples calc && cargo build
```
Comment (snim2) [inline .travis.yml:5]: The example doesn't have its own tests? Would that be helpful documentation to the user?

File: examples/calc/README.md [hunk 2]
```diff
@@ -0,0 +1,11 @@
+# Lexing a simple calculator language
+
+This directory contains a very simple example of a calculator example in
+`lrlex`. Executing `cargo run` processes `src/calc.l` at compile-time;
```
Comment (snim2) [inline examples/calc/README.md]: I think the second sentence here is missing a word?

File: src/lib/build.rs [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+impl<TokId: Copy + Debug + Eq + TypeName> LexerDef<TokId> {
+    pub(crate) fn rust_pp(&self, f: &mut File) -> Result<(), io::Error> {
+        // Header
+        f.write(format!("use lrlex::{{LexerDef, Rule}};
+
+pub fn lexerdef() -> LexerDef<{}> {{
+    let rules = vec![", TokId::type_name()).as_bytes())?;
+
+        // Individual rules
+        for r in self.rules.iter() {
```
Comment (snim2) [inline src/lib/build.rs]: Can say `for r in &self.rules {` here

File: src/lib/build.rs [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+            }
+        }
+    }
+    let inc = read_to_string(&inp).unwrap();
+    let lexerdef = parse_lex::<TokId>(&inc)?;
+    let mut f = File::create(outp)?;
+    let mod_name = inp.as_ref().file_stem().unwrap().to_str().unwrap();
+    f.write(format!("mod {} {{", mod_name).as_bytes())?;
+    lexerdef.rust_pp(&mut f)?;
+    f.write("}".as_bytes())?;
```
Comment (snim2) [inline src/lib/build.rs]: You can use a byte string directly here, `b"}"` rather than calling `.as_bytes()`

File: src/lib/build.rs [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+            f.write(format!("
+Rule::new({}, {}, \"{}\".to_string()).unwrap(),",
+                tok_id, n, r.re_str.replace("\\", "\\\\").replace("\"", "\\\"")).as_bytes())?;
+        }
+
+        // Footer
+        f.write("
+];
+    LexerDef::new(rules)
+}".as_bytes())?;
```
Comment (snim2) [inline src/lib/build.rs]: Likewise, you can use a byte string literal rather than calling `.as_bytes()`

File: src/lib/mod.rs [hunk 6]
```diff
@@ -85,3 +105,32 @@ pub fn build_lex<TokId: Copy + Eq + TryFrom<usize>>(s: &str) -> Result<LexerDef<
...
+                                 outp: Q)
+                              -> Result<(), Box<Error>>
+                           where TokId: Copy + Debug + Eq + TryFrom<usize> + TypeName,
+                                 P: AsRef<Path>,
+                                 Q: AsRef<Path>
+{
+    build::process_file::<TokId, P, Q>(inp, outp)
+}
+
+#[macro_export]
```
Comment (snim2) [inline src/lib/mod.rs]: Should `#[macro_export]` be below the comment?

File: src/lib/mod.rs [hunk 7]
```diff
@@ -85,3 +105,32 @@ pub fn build_lex<TokId: Copy + Eq + TryFrom<usize>>(s: &str) -> Result<LexerDef<
     parse_lex(s)
 }
 
+/// Statically compile all `.l` files in the `src` directory. A file `src/x.l` can then be used in
+/// a crate with `lrlex_mod!(x)`: it exposes a function `lexerdef()` which returns a
+/// [`LexerDef`](struct.LexerDef.html) that can then be used as normal.
+pub fn process_src_dir<TokId: Copy + fmt::Debug + Eq + TryFrom<usize> + TypeName>() -> Result<(), Box<Error>> {
```
Comment (snim2) [inline src/lib/mod.rs:111]: Maybe document that this changes the environment?

File: src/lib/lexer.rs [hunk 8]
```diff
@@ -51,6 +52,23 @@ pub struct Rule<TokId> {
     pub re: Regex
 }
 
+impl<TokId> Rule<TokId> {
+    #[doc(hidden)]
```
Comment (snim2) [inline src/lib/lexer.rs]: Swap `#[doc(hidden)]` with the comment?

**Comment by snim2** [comment]
OK, this generally looks fine to me. I would suggest doing a clippy PR soon-ish because there are a few new syntax features that would tidy up the `lrlex` code a bit, so the output of clippy is quite long now.

File: .travis.yml (hunk 1 above)
Comment (ltratt) [reply .travis.yml:5]: I think the fact that it's just `cargo build` and not `cargo test` is indication enough.

File: examples/calc/README.md (hunk 2 above)
Comment (ltratt) [reply examples/calc/README.md]: Do you mean the "executing" sentence? It reads OK (albeit it's terse) to me, but I'm biased!

File: .travis.yml (hunk 1 above)
Comment (snim2) [reply .travis.yml:5]: Sorry, I meant would adding a test or two help the user understand the API better? Now I've read the whole PR I'm not so sure...

File: src/lib/build.rs (hunk 4 above)
Comment (ltratt) [reply src/lib/build.rs]: Good point.

File: src/lib/build.rs (hunk 3 above)
Comment (ltratt) [reply src/lib/build.rs]: Unfortunately not, as this would move `rules` out of `self` (which is borrowed), so it causes an error.

Comment (snim2) [reply src/lib/build.rs]: It works for me!

File: src/lib/mod.rs (hunk 6 above)
Comment (ltratt) [reply src/lib/mod.rs]: Looking at Rust's library, it looks like it should be. Will change.

File: src/lib/mod.rs (hunk 7 above)
Comment (ltratt) [reply src/lib/mod.rs:111]: I'm confused: it doesn't change the environment (though it does read from the environment).

File: examples/calc/README.md (hunk 2 above)
Comment (snim2) [reply examples/calc/README.md]: No, the one before, and now I read it again I think it has an extra word:
`This directory contains a very simple example of a calculator example in lrlex`.
Should be:
`This directory contains a very simple example of a calculator lexer in lrlex`.
or similar...?

File: src/lib/lexer.rs (hunk 8 above)
Comment (ltratt) [reply src/lib/lexer.rs]: Agreed.

File: src/lib/mod.rs (hunk 7 above)
Comment (snim2) [reply src/lib/mod.rs:111]: Maybe I misunderstood, I thought somewhere you write an env var `OUT_DIR` and then read it again?

File: src/lib/build.rs (hunk 3 above)
Comment (ltratt) [reply src/lib/build.rs]: I've just tried it on a bang-up-to-date nightly and it still dies:

```
   Compiling lrlex v0.1.0 (file:///home/ltratt/lrlex)
error[E0507]: cannot move out of borrowed content
  --> src/lib/build.rs:78:18
   |
78 |         for r in self.rules {
   |                  ^^^^ cannot move out of borrowed content

error: aborting due to previous error

For more information about this error, try `rustc --explain E0507`.
error: Could not compile `lrlex`.
```

Comment (snim2) [reply src/lib/build.rs]: `&self.rules`

File: examples/calc/README.md (hunk 2 above)
Comment (ltratt) [reply examples/calc/README.md]: Agreed.

File: src/lib/build.rs (hunk 3 above)
Comment (ltratt) [reply src/lib/build.rs]: Mea culpa. You're quite correct!

**Comment by ltratt** [comment]
OK, hopefully those commits address things.

**Comment by ltratt** [comment]
[I also agree on doing a Clippy pass, but I'll do that in another PR. I haven't been able to run Clippy on most of these repos for quite a while...]

**Comment by snim2** [comment]
OK, squash away

**Review by snim2 — APPROVED** [review]

**Comment by ltratt** [comment]
Squashed.

**Pull request merged**